// titanium-sdk/lib/titanium.js
/**
 * Titanium SDK Library for Node.js
 * Copyright (c) 2012 by Appcelerator, Inc. All Rights Reserved.
 * Please see the LICENSE file for information about licensing.
 */
var fs = require('fs'),
	path = require('path'),
	async = require('async'),
	appc = require('node-appc'),
	afs = appc.fs,
	version = appc.version,
	manifest = appc.pkginfo.manifest(module),
	platformAliases = {
		// add additional aliases here for new platforms
		'ipad': 'iphone',
		'ios': 'iphone'
	};

exports.module = require('./module');
exports.plugin = require('./plugin');
exports.tiappxml = require('./tiappxml');

exports.manifest = manifest;
exports.availablePlatforms = (manifest.platforms || []).sort();

exports.commonOptions = function (logger, config) {
	return {
		sdk: {
			abbr: 's',
			default: config.app.sdk || 'latest',
			desc: __('Titanium SDK version to use'),
			required: true
		},
		'log-level': {
			callback: function (value) {
				logger.levels[value] && logger.setLevel(value);
			},
			desc: __('minimum logging level'),
			default: config.cli.logLevel || 'warn',
			hint: __('level'),
			values: logger.getLevels()
		}
	};
};

exports.platformOptions = function (logger, config, cli, commandName, finished) {
	var result = {};
	
	if (!commandName) {
		finished(result);
		return;
	}
	
	function set(obj, title, platform) {
		// add the platform and title to the options and flags
		['options', 'flags'].forEach(function (type) {
			if (obj[type]) {
				result[platform] || (result[platform] = {
					platform: platform,
					title: title || platform
				});
				result[platform][type] = obj[type];
			}
		});
	}
	
	// for each platform, fetch their specific flags/options
	async.parallel(manifest.platforms.map(function (platform) {
		return function (callback) {
			var platformDir = path.join(path.dirname(module.filename), '..', '..', '..', platform),
				platformCommand = path.join(platformDir, 'cli', 'commands', '_' + commandName + '.js'),
				command,
				conf,
				title;
			
			if (!afs.exists(platformCommand)) return callback();
			
			command = require(platformCommand);
			if (!command || !command.config) return callback();
			
			// try to get the platform specific configuration
			conf = command.config(logger, config, cli);
			
			try {
				// try to read a title from the platform's package.json
				title = JSON.parse(fs.readFileSync(path.join(platformDir, 'package.json'))).title;
			} catch (e) {}
			
			if (typeof conf == 'function') {
				// async callback
				conf(function (obj) {
					set(obj, title, platform);
					callback();
				});
				return;
			}

			set(conf, title, platform);
			callback();
		};
	}), function () {
		finished(result);
	});
};

exports.validateProjectDir = function(logger, dir) {
	dir = dir || '.';
	
	var d = appc.fs.resolvePath(dir);
	if (!appc.fs.exists(d)) {
		logger.error(__('Project directory does not exist') + '\n');
		process.exit(1);
	}
	
	var tiapp = path.join(d, 'tiapp.xml');
	while (!appc.fs.exists(tiapp) && d != '/') {
		d = path.dirname(d);
		tiapp = path.join(d, 'tiapp.xml');
	}
	
	if (d == '/') {
		logger.error(__('Invalid project directory "%s"', dir) + '\n');
		dir == '.' && logger.log(__("Use the %s property to specify the project's directory", '--dir'.cyan) + '\n');
		process.exit(1);
	}
	
	return d;
};

exports.validateCorrectSDK = function (logger, config, cli, projectDir) {
	// tiapp.xml should exist by the time we get here
	var tiappxml = new exports.tiappxml(path.join(projectDir, 'tiapp.xml')),
		sdk = tiappxml['sdk-version'];
	if (!sdk || version.eq(sdk, manifest.version)) return true;
	
	// check the project's preferred sdk is even installed
	if (sdk == '__global__' || !cli.env.sdks[sdk]) {
		logger.error(__("Unable to compile project because the 'sdk-version' in the tiapp.xml is not installed") + '\n');
		logger.log(__("The project's %s is currently set to %s, which is not installed.", 'sdk-version'.cyan, sdk.cyan) + '\n');
		logger.log(__("Update the %s in the tiapp.xml to one of the installed Titaniums SDKs:", 'sdk-version'.cyan));
		Object.keys(cli.env.sdks).sort().forEach(function (ver) {
			if (ver != '__global__') {
				logger.log('   ' + ver.cyan);
			}
		});
		logger.log(__("or run '%s' to download and install Titanium SDK %s", ('titanium sdk install ' + sdk).cyan, sdk) + '\n');
		process.exit(1);		
	}
	
	var args = cli.argv.$_,
		p = args.indexOf('--sdk');
	if (p != -1) {
		args.splice(p, 2);
	}
	
	var cmd = cli.argv.$ + ' ' + cli.argv.$_.join(' ') + ' --sdk ' + sdk;
	if (version.lt(sdk, '2.2.0')) {
		var platform = platformAliases[cli.argv.platform] || cli.argv.platform;
		cmd = 'python "' + path.join(path.resolve(cli.env.sdks[sdk].path), platform, 'builder.py') + '"';
		if (platform == 'iphone') {
			cmd += ' run "' + projectDir + '"';
		} else if (platform == 'mobileweb') {
			cmd += ' "' + projectDir + '" "' + cli.argv['build-type'] + '"';
		} else if (platform == 'android') {
			// TODO:
		}
	}
	
	// fork or die
	if (config.cli.failOnWrongSDK) {
		logger.error(__('Unable to compile a %s project with Titanium SDK %s', version.format(sdk, 3, 3), version.format(manifest.version, 3, 3)) + '\n');
		logger.log(__("Run '%s' to build using the correct Titanium SDK or update the %s in the tiapp.xml to the current Titaniums SDK: %s",
			cmd.cyan, 'sdk-version'.cyan, manifest.version.cyan) + '\n');
		process.exit(1);
	}
	
	// TODO...
	console.log('Forking correct SDK command: ' + cmd.cyan + '\n');
};

exports.validatePlatformOptions = function (logger, config, cli, commandName) {
	commandName && manifest.platforms.forEach(function (platform) {
		var platformCommand = path.join(path.dirname(module.filename), '..', '..', '..', platform, 'cli', 'commands', '_' + commandName + '.js');
		if (afs.exists(platformCommand)) {
			var command = require(platformCommand);
			command && command.validate && command.validate(logger, config, cli);
		}
	});
};

exports.scrubPlatforms = function (platforms) {
	var r = {
		scrubbed: [], // distinct list of un-aliased platforms
		bad: []
	};
	
	platforms.toLowerCase().split(',').map(function (p) {
		return platformAliases[p] || p;
	}).forEach(function (p) {
		if (manifest.platforms.indexOf(p) == -1) {
			r.bad.push(p);
		} else if (r.scrubbed.indexOf(p) == -1) {
			r.scrubbed.push(p);
		}
	});
	
	return r;
};

exports.validatePlatform = function(logger, platform) {
	var p = platformAliases[platform] || platform;
	if (!p || manifest.platforms.indexOf(p) == -1) {
		logger.error(__('Invalid platform "%s"', platform) + '\n');
		appc.string.suggest(platform, manifest.platforms, logger.log);
		logger.log(__('Available platforms for SDK version %s:', manifest.version));
		manifest.platforms.forEach(function (p) {
			logger.log('    ' + p.cyan);
		});
		logger.log();
		process.exit(1);
	}
	return p;
};

exports.validatePlatforms = function (logger, platforms) {
	var p = platforms;
	// split platforms and make sure each are valid, then return the cleaned up platforms
	return p;
};
