// titanium-sdk/lib/windows.js
/**
 * Detects the Windows Phone development environment and its dependencies.
 *
 * @module lib/wp8
 *
 * @copyright
 * Copyright (c) 2009-2013 by Appcelerator, Inc. All Rights Reserved.
 *
 * @license
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

var appc = require('node-appc'),
	async = require('async'),
	fs = require('fs'),
	path = require('path'),
	__ = appc.i18n(__dirname).__,
	mobilewebPackageJson = {},
	cachedResults;

// need to find the mobile web platform and its package.json
(function findPackageJson(dir) {
	if (dir != '/') {
		var file = path.join(dir, 'mobileweb', 'package.json');
		if (fs.existsSync(file)) {
			mobilewebPackageJson = require(file);
		} else {
			findPackageJson(path.dirname(dir));
		}
	}
}(path.join(__dirname, '..', '..', '..')));

/**
 * Detects current Windows Phone environment.
 * @param {Object} config - The CLI config object
 * @param {Object} opts - Detect options
 * @param {Function} finished - Callback when detection is finished
 */
exports.detect = function detect(config, opts, finished) {
	if (process.platform != 'win32') return finished();

	opts || (opts = {});

	if (cachedResults && !opts.bypassCache) return finished(cachedResults);

	var results = {
		issues: []
	};

	async.parallel({
		'visualstudio': function (next) {
			var keyRegExp = /.+\\(\d+\.\d)_config$/i,
				valueRegExp = /\s(\w+)\s+(REG_\w+)\s+(.+)$/,
				possibleVersions = {},
				vsInfo = {};

			// scan the registry to find some Visual Studio installs
			async.parallel([
				'HKEY_LOCAL_MACHINE\\Software\\Microsoft\\VisualStudio', // there should not be anything here because VS is currently 32-bit and we'll find it next
				'HKEY_LOCAL_MACHINE\\Software\\Wow6432Node\\Microsoft\\VisualStudio', // this is where VS should be found because it's 32-bit
				'HKEY_CURRENT_USER\\Software\\Microsoft\\VisualStudio' // should be the same as the one above, but just to be safe
			].map(function (keyPath) {
				return function (next) {
					appc.subprocess.run('reg', ['query', keyPath], function (code, out, err) {
						if (!code) {
							out.trim().split(/\r\n|\n/).forEach(function (configKey) {
								configKey = configKey.trim();
								var m = configKey.match(keyRegExp);
								if (m) {
									possibleVersions[configKey] = {
										version: m[1],
										configKey: configKey
									};
								}
							});
						}
						next();
					});
				};
			}), function () {
				// if we didn't find any Visual Studios, then we're done
				if (!Object.keys(possibleVersions).length) {
					results.issues.push({
						id: 'WINDOWS_VISUAL_STUDIO_NOT_INSTALLED',
						type: 'error',
						message: __('Microsoft Visual Studio not installed.') + '\n' +
							__('You will be unable to build Windows Phone or Windows Store apps.') + '\n' +
							__('You can install it from %s.', '__http://appcelerator.com/visualstudio__')
					});
					return next();
				}

				// fetch Visual Studio install information
				async.parallel(Object.keys(possibleVersions).map(function (configKey) {
					return function (next) {
						appc.subprocess.run('reg', ['query', configKey, '/v', '*'], function (code, out, err) {
							if (!code) {
								var ver = possibleVersions[configKey].version,
									info = vsInfo[ver] = {
										version: ver,
										registryKey: configKey,
										supported: appc.version.satisfies(ver, mobilewebPackageJson.vendorDependencies['visual studio'], true),
										vcvarsall: null,
										wpsdk: null,
										selected: false
									};

								// get only the values we are interested in
								out.trim().split(/\r\n|\n/).forEach(function (line) {
									var parts = line.trim().split('   ').map(function (p) { return p.trim(); });
									if (parts.length == 3) {
										if (parts[0] == 'CLR Version') {
											info.clrVersion = parts[2];
										} else if (parts[0] == 'ShellFolder') {
											info.path = parts[2];
										}
									}
								});

								// verify that this Visual Studio actually exists
								if (info.path && fs.existsSync(info.path) && fs.existsSync(path.join(info.path, 'Common7', 'IDE', 'devenv.exe'))) {
									// get the vcvarsall script
									var vcvarsall = path.join(info.path, 'VC', 'vcvarsall.bat');
									if (fs.existsSync(vcvarsall)) {
										info.vcvarsall = vcvarsall;
									}

									// detect all Windows Phone SDKs
									var wpsdkDir = path.join(info.path, 'VC', 'WPSDK');
									fs.existsSync(wpsdkDir) && fs.readdirSync(wpsdkDir).forEach(function (ver) {
										var vcvarsphone = path.join(wpsdkDir, ver, 'vcvarsphoneall.bat');
										if (fs.existsSync(vcvarsphone) && /^wp\d+$/i.test(ver)) {
											// we found a windows phone sdk!
											var name = (parseInt(ver.replace(/^wp/i, '')) / 10).toFixed(1);
											info.wpsdk || (info.wpsdk = {});
											info.wpsdk[name] = {
												vcvarsphone: vcvarsphone
											};
										}
									});
								}
							}

							if (info.vcvarsall) {
								appc.subprocess.getRealName(info.vcvarsall, function (err, file) {
									if (!err) {
										info.vcvarsall = file
									}
									next();
								});
							} else {
								next();
							}
						});
					};
				}), function () {
					// double check if we didn't find any Visual Studios, then we're done
					if (!Object.keys(vsInfo).length) {
						results.issues.push({
							id: 'WINDOWS_VISUAL_STUDIO_NOT_INSTALLED',
							type: 'error',
							message: __('Microsoft Visual Studio not installed.') + '\n' +
								__('You will be unable to build Windows Phone or Windows Store apps.') + '\n' +
								__('You can install it from %s.', '__http://appcelerator.com/visualstudio__')
						});
						return next();
					}

					if (Object.keys(vsInfo).every(function (v) { return !vsInfo[v].wpsdk; })) {
						results.issues.push({
							id: 'WINDOWS_PHONE_SDK_NOT_INSTALLED',
							type: 'warning',
							message: __('Microsoft Windows Phone SDK not installed.') + '\n' +
								__('You will be unable to build Windows Phone apps.') + '\n' +
								__('You can install it from %s.', '__http://appcelerator.com/windowsphone__')
						});
						return next();
					}

					var preferredVersion = config.get('windows.visualstudio.selectedVersion');
					vsInfo[preferredVersion && vsInfo[preferredVersion] ? preferredVersion : Object.keys(vsInfo).sort().pop()].selected = true;

					next(null, vsInfo);
				});
			});
		},

		'windowsphone': function (next) {
			var wpInfo = {};

			async.parallel([
				'HKEY_LOCAL_MACHINE\\Software\\Microsoft\\Microsoft SDKs\\WindowsPhone', // probably nothing here
				'HKEY_LOCAL_MACHINE\\Software\\Wow6432Node\\Microsoft\\Microsoft SDKs\\WindowsPhone' // this is most likely where WPSDK will be found
			].map(function (keyPath) {
				return function (next) {
					appc.subprocess.run('reg', ['query', keyPath], function (code, out, err) {
						var keyRegExp = /.+\\(v\d+\.\d)$/;
						if (!code) {
							out.trim().split(/\r\n|\n/).forEach(function (key) {
								key = key.trim();
								var m = key.match(keyRegExp),
									version = m[1].replace(/^v/, '');
								if (m) {
									wpInfo[version] = {
										version: version,
										registryKey: keyPath + '\\' + m[1],
										supported: appc.version.satisfies(version, mobilewebPackageJson.vendorDependencies['windows phone sdk'], true),
										path: null,
										xapDeployCmd: null,
										selected: false
									};
								}
							});
						}
						next();
					});
				};
			}), function () {
				// check if we didn't find any Windows Phone SDKs, then we're done
				if (!Object.keys(wpInfo).length) {
					results.issues.push({
						id: 'WINDOWS_PHONE_SDK_NOT_INSTALLED',
						type: 'error',
						message: __('Microsoft Windows Phone SDK not installed.') + '\n' +
							__('You will be unable to build Windows Phone apps.') + '\n' +
							__('You can install it from %s.', '__http://appcelerator.com/windowsphone__')
					});
					return next();
				}

				// fetch Windows Phone SDK install information
				async.parallel(Object.keys(wpInfo).map(function (ver) {
					return function (next) {
						appc.subprocess.run('reg', ['query', wpInfo[ver].registryKey + '\\Install Path', '/v', '*'], function (code, out, err) {
							if (!code) {
								// get only the values we are interested in
								out.trim().split(/\r\n|\n/).forEach(function (line) {
									var parts = line.trim().split('   ').map(function (p) { return p.trim(); });
									if (parts.length == 3) {
										if (parts[0] == 'Install Path') {
											wpInfo[ver].path = parts[2];
											var xapDeployCmd = path.join(parts[2], 'Tools', 'XAP Deployment', 'XapDeployCmd.exe');
											if (fs.existsSync(xapDeployCmd)) {
												wpInfo[ver].xapDeployCmd = xapDeployCmd;
											}
										}
									}
								});
							}
							next();
						});
					};
				}), function () {
					// double check if we didn't find any Windows Phone SDKs, then we're done
					if (Object.keys(wpInfo).every(function (v) { return !wpInfo[v].path; })) {
						results.issues.push({
							id: 'WINDOWS_PHONE_SDK_NOT_INSTALLED',
							type: 'error',
							message: __('Microsoft Windows Phone SDK not installed.') + '\n' +
								__('You will be unable to build Windows Phone apps.') + '\n' +
								__('You can install it from %s.', '__http://appcelerator.com/windowsphone__')
						});
						return next();
					}

					if (Object.keys(wpInfo).every(function (v) { return !wpInfo[v].xapDeployCmd; })) {
						results.issues.push({
							id: 'WINDOWS_PHONE_SDK_MISSING_XAP_DEPLOY_CMD',
							type: 'error',
							message: __('Microsoft Windows Phone SDK is missing the XAP deploy command.') + '\n' +
								__('You will be unable to build Windows Phone apps.') + '\n' +
								__('You can install it from %s.', '__http://appcelerator.com/windowsphone__')
						});
						return next();
					}

					var preferredVersion = config.get('windows.wpsdk.selectedVersion');
					wpInfo[preferredVersion && vsInfo[preferredVersion] ? preferredVersion : Object.keys(wpInfo).sort().pop()].selected = true;

					next(null, wpInfo);
				});
			});
		}
	}, function (err, result) {
		appc.util.mix(results, result);

		var selectedVisualStudio = exports.getSelectedVisualStudio(results),
			selectedWindowsPhoneSDK = exports.getSelectedWindowsPhoneSDK(results);

		async.parallel({
			'msbuild': function (next) {
				// now that we've detected Visual Studio, we can try to find MSBuild
				if (!selectedVisualStudio) return next();

				appc.subprocess.run('cmd', [ '/C', selectedVisualStudio.vcvarsall + ' && MSBuild /version' ], function (code, out, err) {
					var msbuildInfo = null;

					if (code) {
						results.issues.push({
							id: 'WINDOWS_MSBUILD_ERROR',
							type: 'error',
							message: __('Failed to run MSBuild.') + '\n' +
								__('This is most likely due to Visual Studio cannot find a suitable .NET framework.') + '\n' +
								__('Please install the latest .NET framework.')
						});
					} else {
						var chunks = out.trim().split(/\r\n\r\n|\n\n/);
						chunks.shift(); // strip off the first chunk

						msbuildInfo = {
							version: chunks.shift().split(/\r\n|\n/).pop().trim()
						};

						if (!appc.version.satisfies(msbuildInfo.version, mobilewebPackageJson.vendorDependencies['msbuild'])) {
							results.issues.push({
								id: 'WINDOWS_MSBUILD_TOO_OLD',
								type: 'error',
								message: __('The MSBuild version %s is too old.', msbuildInfo.version) + '\n' +
									__("Titanium requires .NET MSBuild '%s'.", mobilewebPackageJson.vendorDependencies['msbuild']) + '\n' +
									__('Please install the latest .NET framework.')
							});
						}
					}

					next(null, msbuildInfo);
				});
			},

			'devices': function (next) {
				if (selectedWindowsPhoneSDK && selectedWindowsPhoneSDK.xapDeployCmd) {
					appc.subprocess.run(selectedWindowsPhoneSDK.xapDeployCmd, '/EnumerateDevices', function (code, out, err) {
						var devices = null;

						if (err) {
							results.issues.push({
								id: 'WINDOWS_PHONE_ENUMERATE_DEVICES_FAILED',
								type: 'error',
								message: __('Failed to enumerate Windows Phone devices.') + '\n' +
									__('Ensure that the Windows Phone SDK is properly installed.')
							});
						} else {
							var deviceRegExp = /^ ([0-9]*)\t\t(.*)$/;
							devices = {};
							out.trim().split(/\r\n|\n/).forEach(function (line) {
								var m = line.match(deviceRegExp);
								if (m) {
									devices[m[1]] = m[2];
								}
							});
						}

						next(null, devices);
					});
				} else {
					next();
				}
			}
		}, function (err, result) {
			finished(cachedResults = appc.util.mix(results, result));
		});
	});
};

exports.getSelectedVisualStudio = function getSelectedVisualStudio(env) {
	if (env.visualstudio) {
		return env.visualstudio[Object.keys(env.visualstudio).filter(function (v) {
			return env.visualstudio[v].selected;
		}).pop()];
	}
};

exports.getSelectedWindowsPhoneSDK = function getSelectedWindowsPhoneSDK(env) {
	if (env.windowsphone) {
		return env.windowsphone[Object.keys(env.windowsphone).filter(function (v) {
			return env.windowsphone[v].selected;
		}).pop()];
	}
};
